generators/Exercise.cs

```diff
@@ -1,23 +1,74 @@
-﻿using Generators.Input;
+﻿using System.Collections.Generic;
+using System.Linq;
+using Generators.Input;
 using Generators.Output;
 
 namespace Generators
 {
     public abstract class Exercise
     {
-        protected Exercise()
+        private static readonly ExerciseWriter ExerciseWriter = new ExerciseWriter();
+        
+        public string Name => GetType().ToExerciseName();
+
+        protected CanonicalData CanonicalData { get; private set; }
+        
+        public void Regenerate(CanonicalData canonicalData)
+        {
+            CanonicalData = canonicalData;
+            UpdateCanonicalData(CanonicalData);
+
+            ExerciseWriter.WriteToFile(this);
+        }
+
+        protected virtual void UpdateCanonicalData(CanonicalData canonicalData)
+        {
+        }
+
+        public string Render() => RenderTestClass();
+
+        protected virtual string RenderTestClass() => CreateTestClass().Render();
+
+        protected virtual TestClass CreateTestClass() => new TestClass
         {
-            Name = GetType().ToExerciseName();
-            CanonicalData = CanonicalDataParser.Parse(Name);
+            ClassName = Name.ToTestClassName(),
+            Methods = RenderTestMethods(),
+            CanonicalDataVersion = CanonicalData.Version,
+            UsingNamespaces = GetUsingNamespaces()
+        };
+
+        private HashSet<string> GetUsingNamespaces()
+        {
+            var usingNamespaces = new HashSet<string> { "Xunit" };
+
+            foreach (var canonicalDataCase in CanonicalData.Cases.Where(canonicalDataCase => canonicalDataCase.ExceptionThrown != null))
```

Comment:
This could be simplified to:
```
if (CanonicalData.Cases.Any(canonicalDataCase => canonicalDataCase.ExceptionThrown != null))
    usingNamespaces.Add(canonicalDataCase.ExceptionThrown.Namespace);
```